REST API integration example (#212)
3scale/zync: examples/rest-api/README.md

# Zync REST API example

This example project implements Zync's REST API protocol to synchronize OAuth2 clients.

## Prerequisites

> Given 3scale API is configured to use:

Reviewer: ```suggestion
Ensure 3scale API is configured to use:
```

>   * OpenID Connect as the Authentication,

Reviewer: ```suggestion
  * Authentication uses OpenID Connect.
```

>   * "REST API" as a OpenID Connect Issuer Type and

Reviewer: ```suggestion
  * OpenID Connect Issuer Type is set to "REST API".
```

>   * "http://id:secret@example.com/api" as OpenID Connect Issuer.

Reviewer: ```suggestion
  * OpenID Connect Issuer is set to `http://id:secret@example.com/api`.
```

> When a 3scale application is created/updated/deleted zync will try to replay that change to "http://example.com/api".

Reviewer: ```suggestion
When you create, update or delete a 3scale application, Zync attempts to replay the change to "http://example.com/api".
```
_Zync_ is a product name, hence capitalized, unless the original name remains with small case.

## Creating, updating and deleting Clients

> Zync will make following requests to create/update/delete clients:

Reviewer: ```suggestion
Zync sends the following requests to create, update or delete clients:
```

* `PUT /clients/:client_id` (create, update)
* `DELETE /clients/:client_id` (delete)

> All endpoints must reply 2xx status code. Otherwise the request will be retried.

Reviewer 2: @mikz is there any limit on number of period of retries? 

What if the DELETE responds with `404`? I believe it can cause an infinite loop of trying to delete the non-existing client, e.g. https://issues.jboss.org/browse/THREESCALE-2206

Author: Then it should respond with 200 that the client is gone. It has exponential backoff and yes, there is some limit. It is the same for every object.

Reviewer 2: @mikz Can you please point me to the place where this is configured/set?

Author: I believe that for most tasks we are using the default https://github.com/chanks/que/blob/master/docs/error_handling.md#error-handling

Reviewer 2: @mikz thanks for the info!

> ### Payload

Reviewer: ```suggestion
### Configuring the payload
```

> The request payload in case of create and update is `application/json`:

Reviewer: ```suggestion
When creating or updating, the payload request is the following `application/json`:
```

```json
{
  "client_id": "ee305610",
  "client_secret": "ac0e42db426b4377096c6590e2b06aed",
  "client_name": "oidc-app",
  "redirect_uris": ["http://example.com"],
  "grant_types": ["client_credentials", "password"]
}
```

The request to delete a client has no payload.

## Using OAuth2 authentication

> Zync will make GET request to `/.well-known/openid-configuration` endpoint and expect an `application/json` response.

Reviewer: ```suggestion
Zync sends GET requests to `/.well-known/openid-configuration` endpoint, and expects an `application/json` response.
```

The response payload should contain following: 

```json
{
  "token_endpoint": "http://idp.example.com/auth/realm/token"
}
```

> Zync will use that `token_endpoint` URL to exchange the client_id and client_secret provided in the OpenID Connect Issuer URL 

Reviewer: ```suggestion
Zync uses the `token_endpoint` URL to exchange the _client_id_ and _client_secret_ provided in the OpenID Connect Issuer URL.
```

for an access token using the OAuth2 protocol.

> If the API responds with not successful response, Zync will fallback to HTTP Basic/Digest authentication using provided credentials.

Reviewer: ```suggestion
If the API responds with a not successful response, Zync uses HTTP Basic/Digest authentication using the provided credentials.
```
